File: AGRA_0.5/Hw04/Ejemplos/kosaraju2.cpp
/*
Algoritmo de Kosaraju (SCC)
Autor: Carlos Ramirez
Fecha: Marzo 23 de 2020

En esta version se utiliza un vector en el cual se van agregando
los elementos al final mientras se realiza el primer recorrido.
De esta manera en el vector ord se almacena el orden topologico
del grafo al reves. Ademas, se utiliza una variable que almacena
la cantidad de SCC y otra que almacena los nodos en cada SCC.
 */

#include <vector>
#include <list>
#include <iostream>

using namespace std;

int n, numSCC;
vector<vector<int> > adj(5000);
vector<vector<int> > adjT(5000);
bool visitados[5000];
vector<int> ord;
int sccInd[5000];
vector<vector<int> > sccNodos;

void kosarajuAux(int);
void asignar(int, int);

void kosaraju(){
  int i;

  // primer DFS --> O(n + m)
  for(i = 0; i < n; i++)
    kosarajuAux(i);

  numSCC = 0;
  // O(n)
  for(i = 0; i < n; i++)
    sccInd[i] = -1;

  // segundo DFS --> (n + m)
  for(i = n - 1; i >= 0; i--){
    if(sccInd[ord[i]] == -1){
      numSCC++;
      sccNodos.push_back(vector<int>());
      asignar(ord[i], numSCC - 1);
    }
  }
}

void kosarajuAux(int v){
  if(!visitados[v]){
    visitados[v] = true;
    
    for(int i = 0; i < adj[v].size(); i++)
      kosarajuAux(adj[v][i]);

    ord.push_back(v);
  }
}

void asignar(int u, int num){
  sccInd[u] = num;
  sccNodos[numSCC - 1].push_back(u);

  for(int i = 0; i < adjT[u].size(); i++){
    if(sccInd[adjT[u][i]] == -1)
      asignar(adjT[u][i], num);
  }
}

int main(){
  int m, i, j, aux1, aux2;

  cin >> n >> m;

  for(i = 0; i < m; i++){
    cin >> aux1 >> aux2;
    adj[aux1].push_back(aux2);
    adjT[aux2].push_back(aux1);
  }

  kosaraju();

  cout << "Total SCC: " << numSCC << endl;

  for(i = 0; i < numSCC; i++){
    cout << "SCC " << i + 1 << ": ";
    for(j = 0; j < sccNodos[i].size(); j++)
      cout << sccNodos[i][j] << " ";
    cout << endl;
  }
  
  for(i = 0; i < n; i++){
    cout << sccInd[i] << " ";
  }

  cout << endl;
  
  return 0;
}

File: AGRA/Hw04/Ejemplos/kosaraju2.py
from sys import stdin
from collections import deque

MAX = 10000
adj, adjT = [[] for i in range(MAX)], [[] for i in range(MAX)]
visitados, sccInd = [False for i in range(MAX)], [-1 for i in range(MAX)]
n, numSCC = int(), 0
sccNodos, ord = [], deque()

# Complejidad Temporal : O(n + m)
# Complejidad Espacial : O(n)
def kosaraju():
    global numSCC
    
    # primer DFS --> O(n + m)
    for i in range(n): kosarajuAux(i)

    #O(n)
    for i in range(n): sccInd[i] = -1

    #segundo DFS --> O(n + m)
    numSCC = 0
    for i in range(n):
        if sccInd[ord[i]] == -1:
            numSCC += 1
            sccNodos.append([])
            asignar(ord[i], numSCC)

def kosarajuAux(v):
    if not visitados[v]:
        visitados[v] = True

        for i in range(len(adj[v])):
            kosarajuAux(adj[v][i])

        ord.appendleft(v)

def asignar(u, num):
    sccInd[u] = num
    sccNodos[num - 1].append(u)

    for i in range(len(adjT[u])):
        if sccInd[adjT[u][i]] == -1:
            asignar(adjT[u][i], num)

def main():
    global n
    n, m = list(map(int, stdin.readline().split()))

    for i in range(m):
        u, v = list(map(int, stdin.readline().split()))
        adj[u].append(v)
        adjT[v].append(u)

    print("Grafo")
    for i in range(n):
        print("Nodo %d:" % i)
        for j in range(len(adj[i])):
            print("%d" % adj[i][j], end = ' ')
        print("")

    kosaraju()
    print("Total SCC: %d" % numSCC)

    for i in range(numSCC):
        print("SCC %d: "% (i + 1), end = '')
        print(sccNodos[i])

    for i in range(n):
        print("%d " % sccInd[i], end = '')
    print("")

main()

    
    
